src/content/en/fundamentals/media/fast-playback-with-video-preload.md

```diff
@@ -0,0 +1,571 @@
+  return cache.match(videoFileUrl)
+  .then(cacheResponse => {
+    // Let's return cached response if video is already in the cache.
+    if (cacheResponse) {
+      return cacheResponse;
+    }
+    // Otherwise, fetch the video from the network.
+    return fetch(videoFileUrl)
+    .then(networkResponse => {
+      // Add the response to the cache and return network response in parallel.
+      cache.put(videoFileUrl, networkResponse.clone());
+      return networkResponse;
+    });
+  });
+}
+```
+
+Note that if were to use HTTP Range request, I would have to recreate manually
+a `Response` object as the Cache API doesn't support Range responses [yet]. Be
+mindful that calling `networkResponse.arrayBuffer()` fetches the whole content
+of the response at once into renderer memory, hence why you may want to use
+small ranges.
+
+For reference, here's the modified part of the code above to save HTTP Range
+requests to the video pre-cache.
+
+```
+    ...
+    return fetch(videoFileUrl, { headers: { range: 'bytes=0-567139' } })
+    .then(networkResponse => networkResponse.arrayBuffer())
+    .then(data => {
+      const response = new Response(data);
+      // Add the response to the cache and return network response in parallel.
+      cache.put(videoFileUrl, response.clone());
+      return response;
+    });
+```
+
+#### Play video
+
+When user clicks a play button, we'll fetch first segment of video available in
+the Cache API so that playback starts immediately if available. Otherwise,
+we'll simply fetch it from the network. Keep in mind that browser and users may
+decide to clear the [Cache].
+
+As seen before, we use MSE to feed that first segment of video to the video
+element.
+
+```
+function onPlayButtonClick(videoFileUrl) {
+  video.load(); // Used to be able to play video later.
+
+  caches.open('video-pre-cache')
+  .then(cache => fetchAndCache(videoFileUrl, cache)) // Defined above.
+  .then(response => response.arrayBuffer())
+  .then(data => {
+    const mediaSource = new MediaSource();
+    video.src = URL.createObjectURL(mediaSource);
+    mediaSource.addEventListener('sourceopen', sourceOpen);
+
+    function sourceOpen() {
+      URL.revokeObjectURL(video.src);
+
+      const sourceBuffer = mediaSource.addSourceBuffer('video/webm; codecs="vp9"');
+      sourceBuffer.appendBuffer(data);
+      sourceBuffer.addEventListener('updateend', updateEnd, { once: true });
+    }
+
+    function updateEnd() {
+      video.play()
+      .then(_ => {
+        // TODO: Fetch the rest of the video when user starts playing video.
+      });
+    }
+  });
+}
+```
+
+Warning: For cross-origin resources, make sure your CORS headers are set
+properly as opaque responses retrieved with `fetch(videoFileUrl, { mode:
+'no-cors' })` are not allowed to feed any video or audio element.
+
+### Create Range responses with a Service Worker
+
+Now what if you fetched an entire video file and saved it in the Cache API. When
```

Comment:
I'd add a quick intro explaining "If you use a Service Worker and the Cache API, you have complete control over when and how much data is cached."